runtime/Java/src/org/antlr/v4/runtime/tree/TreeRewriteSupport.java

```diff
@@ -0,0 +1,157 @@
+ *  override two key methods:
+ * <pre>
+	 public class MyTreeRewriter extends MyParserNameBaseVisitor<ParseTree> {
+	    @Override
+	    public ParseTree visitChildren(RuleNode node) {
+	        return TreeRewriteSupport.rewriteChildren(this, node);
+	    }
+
+	    @Override
+	    public ParseTree visitTerminal(TerminalNode node) {
+	        return node;
+	    }
+		... any/all of your visitX() methods for rules in MyParserName ...
+	 }
+ </pre>
+ *
+ *  See also TestTestRewriteVisitors.
+ *
+ *  The visitor event methods return either the same node, if no change,
+ *  or a node/subtree to replace the event method parameter. Null means
+ *  delete that subtree/leaf.
+ *
+ *  The tree returned from the overall visit() method will be a
+ *  maximally-sharing copy of the original. I.e., it treats the original
+ *  tree as immutable.  Replacing a child node of t will force a copy of t
+ *  so its children list can be modified w/o altering original tree. That
+ *  implies a copy is made of all nodes on the path from t to the root
+ *  of the tree (the spine).
+ *
+ *  The start and stop token parse tree fields are not altered in any way.
+ *  They are meaningless after altering the tree.  Use node's getText()
+ *  to get text of rewrite trees, not getText(RuleContext) from token stream,
+ *  which relies on node start/stop fields.
+ *
+ *  @since 4.6.1
+ */
+public class TreeRewriteSupport {
+	public static ParseTree rewriteChildren(AbstractParseTreeVisitor<ParseTree> visitor, RuleNode tree) {
+		ParserRuleContext resultTree = (ParserRuleContext)tree;
+		int n = tree.getChildCount();
+		int nextChildToAlter = 0; // because of deletions, next child might not be i
+		for (int i=0; i<n; i++) {
+			ParseTree c = tree.getChild(i);
+			// visit child so we can make substitutions in child before replacing it as child of this node
+			ParseTree replacement = visitor.visit(c);
+			if ( replacement != c ) {
+				// To alter child list, we make copy of this node and set child in the copy
+				if ( resultTree==tree ) { // Only make one copy
+					resultTree = copy((ParserRuleContext)tree);
+				}
+				if ( replacement!=null ) {
+					_setChild(resultTree, nextChildToAlter, replacement);
+				}
+				else {
+					resultTree.children.remove(nextChildToAlter);
+					nextChildToAlter--; // to delete, we adjust so nextChildToAlter doesn't advance
+				}
+			}
+			nextChildToAlter++; // track i unless deletion
+		}
+
+		return resultTree; // return original or possibly copy of original
+	}
+
+	/** Hook replacement in as tree's ith child, set parent of new child properly.
+	 *  If replacement is a token node, make copy of it before setting parent
+	 *  if not a clone already (parent is null).
+	 *
+	 *  This is not meant as general API for users. It is destructive to tree
+	 *  argument, which we assume is a copy made from original incoming tree.
+	 */
+	protected static void _setChild(RuleNode tree, int i, ParseTree replacement) {
```

Comment:
I think they could be handy and they are needed to alter child list at a specific subtree.